### PR TITLE
03 webhook invocation

### PR DESCRIPTION
**Pull request by layko-andrey**

### CODE REVIEW

**Review by trenatov — DISMISSED** [review]

File: build.gradle [hunk 1]
```diff
@@ -36,6 +36,9 @@ dependencies {
     implementation('org.web3j:core:4.6.3')
     implementation('com.squareup.okhttp3:okhttp:4.8.1')
 
+    // Reactive
+    implementation('io.projectreactor:reactor-core')
```
Comment (ruslanys) [inline build.gradle]: It's already in-place

File: src/main/kotlin/io/openfuture/state/config/RedisConfig.kt [hunk 2]
```diff
@@ -17,5 +17,4 @@ class RedisConfig {
         val serializationContext = RedisSerializationContext.newSerializationContext<String, Any>(serializer).build()
         return ReactiveRedisTemplate(factory, serializationContext)
     }
-
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/config/RedisConfig.kt:23]: Do not remove the empty line

File: src/main/kotlin/io/openfuture/state/domain/TransactionQueueTask.kt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package io.openfuture.state.domain
+
+data class TransactionQueueTask(
+        val transactionId: String = ""
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/domain/TransactionQueueTask.kt]: Styleguide?

Comment (ruslanys) [reply src/main/kotlin/io/openfuture/state/domain/TransactionQueueTask.kt]: remove default value

File: src/main/kotlin/io/openfuture/state/extensions/Extensions.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+fun ReactiveRedisTemplate<String, Any>.keyToByteBuffer(key: String): ByteBuffer {
+    return this.serializationContext.keySerializationPair.write(key)
+}
+
+fun ReactiveRedisTemplate<String, Any>.valueToByteBuffer(value: Any): ByteBuffer {
+    return this.serializationContext.valueSerializationPair.write(value)
+}
+
+fun LocalDateTime.toMillisDouble(): Double {
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/extensions/Extensions.kt]: Remove

File: src/main/kotlin/io/openfuture/state/repository/WalletQueueRedisRepository.kt [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                    listOf(redisTemplate.valueToByteBuffer(transaction))
+            )
+
+            Flux.zip(walletAdd, transactionAdd)
+        }
+
+        result.awaitLast()
+    }
+
+    suspend fun score(walletId: String): Mono<Double> {
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/repository/WalletQueueRedisRepository.kt]: Replace function. `isExists`?

Comment (ruslanys) [reply src/main/kotlin/io/openfuture/state/repository/WalletQueueRedisRepository.kt]: suspend fun <> Mono

File: src/main/kotlin/io/openfuture/state/repository/WalletQueueRedisRepository.kt [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    suspend fun lock(walletId: String): Boolean {
+        return locks.setIfAbsentAndAwait(
+                "LOCK:${walletId}",
+                LocalDateTime.now(),
+                webhookProperties.lockTTL
+        )
+    }
+
+    suspend fun walletsScheduledTo(toDate: LocalDateTime): Flux<String> {
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/repository/WalletQueueRedisRepository.kt]: Flow

Comment (ruslanys) [reply src/main/kotlin/io/openfuture/state/repository/WalletQueueRedisRepository.kt]: !suspend

File: src/main/kotlin/io/openfuture/state/service/DefaultWalletQueueService.kt [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+import io.openfuture.state.domain.TransactionQueueTask
+import io.openfuture.state.extensions.toMillisDouble
+import io.openfuture.state.repository.WalletQueueRedisRepository
+import kotlinx.coroutines.reactive.awaitSingle
+import org.springframework.stereotype.Service
+import java.time.LocalDateTime
+
+@Service
+class DefaultWalletQueueService(
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/service/DefaultWalletQueueService.kt]: Refactor

File: src/main/kotlin/io/openfuture/state/service/DefaultWebhookService.kt [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+import io.openfuture.state.domain.Transaction
+import io.openfuture.state.domain.TransactionQueueTask
+import io.openfuture.state.domain.Wallet
+import org.springframework.stereotype.Service
+import java.time.LocalDateTime
+
+@Service
+class DefaultWebhookService(
+        private val walletQueueService: WalletQueueService,
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/service/DefaultWebhookService.kt]: Remove service layer for this purpose.

Comment (ruslanys) [reply src/main/kotlin/io/openfuture/state/service/DefaultWebhookService.kt]: Merge these services into one redis repo

File: src/main/kotlin/io/openfuture/state/service/DefaultWebhookService.kt [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+        val score = walletQueueService.score(wallet.id)
+        if (score == null) {
+            walletQueueService.add(wallet.id, transactionTask)
+        } else {
+            transactionsQueueService.add(wallet.id, transactionTask)
+        }
+    }
+
+    override suspend fun walletsScheduledForNow(): List<String> {
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/service/DefaultWebhookService.kt]: Refactor.

1. Only one wallet
2. Method should check that score is less than now()
3. Return Wallet + Score
4. Optional argument score

File: src/main/kotlin/io/openfuture/state/webhhok/DefaultWebhookExecutor.kt [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package io.openfuture.state.webhhok
+
+import org.springframework.stereotype.Service
+
+@Service
+class DefaultWebhookExecutor: WebhookExecutor {
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/webhhok/DefaultWebhookExecutor.kt]: We cannot merge it

File: src/main/kotlin/io/openfuture/state/webhhok/WebhookProcessor.kt [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+*/
+
+@Component
+class WebhookProcessor(
+        private val webhookService: WebhookService,
+        private val webhookExecutor: WebhookExecutor
+) {
+
+    @Scheduled(fixedDelayString = "#{@webhookInvocationProcessDelay}", initialDelay = 1000)
+    fun process() = GlobalScope.launch {
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/webhhok/WebhookProcessor.kt]: Replace `GlobalScope.launch` with `runBlocking`

**Review by ruslanys — CHANGES_REQUESTED** [review]

File: src/main/kotlin/io/openfuture/state/extensions/Extensions.kt (hunk 4 above)
Comment (layko-andrey) [reply src/main/kotlin/io/openfuture/state/extensions/Extensions.kt]: Done

File: src/main/kotlin/io/openfuture/state/webhhok/WebhookProcessor.kt (hunk 11 above)
Comment (layko-andrey) [reply src/main/kotlin/io/openfuture/state/webhhok/WebhookProcessor.kt]: Done

File: src/main/kotlin/io/openfuture/state/webhhok/DefaultWebhookExecutor.kt (hunk 10 above)
Comment (layko-andrey) [reply src/main/kotlin/io/openfuture/state/webhhok/DefaultWebhookExecutor.kt]: Removed

File: src/main/kotlin/io/openfuture/state/service/DefaultWebhookService.kt (hunk 8 above)
Comment (layko-andrey) [reply src/main/kotlin/io/openfuture/state/service/DefaultWebhookService.kt]: Done

File: src/main/kotlin/io/openfuture/state/domain/TransactionQueueTask.kt (hunk 3 above)
Comment (layko-andrey) [reply src/main/kotlin/io/openfuture/state/domain/TransactionQueueTask.kt]: Done

File: src/main/kotlin/io/openfuture/state/repository/WalletQueueRedisRepository.kt (hunk 5 above)
Comment (layko-andrey) [reply src/main/kotlin/io/openfuture/state/repository/WalletQueueRedisRepository.kt]: Renamed to walletScode(). Returned valuewould be used for re-scheduling of wallet later

File: src/main/kotlin/io/openfuture/state/service/DefaultWebhookService.kt (hunk 9 above)
Comment (layko-andrey) [reply src/main/kotlin/io/openfuture/state/service/DefaultWebhookService.kt]: Done

**Review by trenatov — APPROVED** [review]

File: src/main/kotlin/io/openfuture/state/domain/WalletIdentity.kt [hunk 12]
```diff
@@ -1,6 +1,6 @@
 package io.openfuture.state.domain
 
 data class WalletIdentity(
-        val blockchain: String,
-        val address: String
+    val blockchain: String,
+    val address: String
 )
```
Comment (ruslanys) [inline src/main/kotlin/io/openfuture/state/domain/WalletIdentity.kt:6]: now new line

**Review by ruslanys — APPROVED** [review]

**Pull request merged**